src/Agent.Worker/ContainerOperationProvider.cs

```diff
@@ -142,6 +147,74 @@ public async Task StopContainersAsync(IExecutionContext executionContext, object
             await RemoveContainerNetworkAsync(executionContext, _containerNetwork);
         }
 
+        private async Task<string> GetMSILoginToken(CancellationToken cancellationToken)
+        {
+            // Check environment variable for debugging
+            var envVar = System.Environment.GetEnvironmentVariable("DEBUG_MSI_LOGIN_INFO");
+            // Future: Set this client id. This is the MSI client ID.
+            TokenCredential credential = new ManagedIdentityCredential(clientId: null);
+            if (envVar == "1")
+            {
+                // Future: Set this client id. This is the MSI client ID.
+                credential = new ChainedTokenCredential(new ManagedIdentityCredential(clientId: null), new VisualStudioCredential(), new AzureCliCredential());
+            }
+            var accessToken = await credential.GetTokenAsync(new TokenRequestContext(new[] { "https://management.core.windows.net/" }), cancellationToken);
+            // To print the token, you can convert it to string 
+            return accessToken.Token.ToString();
+        }
+
+        private async Task<string> GetACRPasswordFromAADToken(string AADToken, string tenantId, string registryServer, int retryCount, int timeToWait, CancellationToken cancellationToken)
+        {
+            Uri url = new Uri(registryServer + "/oauth2/exchange");
+            const int retryLimit = 5;
+            using var httpClientHandler = HostContext.CreateHttpClientHandler();
```

Comment:
Can you please educate me why we need to pass Custome HttpClientHandler and what it does ?

Reply:
Ran into some errors when not using a custom HttpClientHandler. Will try to see if it works without a custom HttpClientHandler.

Reply:
L0 test failed without creating a custom HttpClientHander:
The following code is using Raw HttpClient() which will not follow the proxy setting agent have. Please use New HttpClient(HostContext.CreateHttpClientHandler())